src/setuptools_scm/discover.py

```diff
@@ -2,15 +2,40 @@
 from .utils import trace, iter_entry_points
 
 
-def iter_matching_entrypoints(path, entrypoint, search_parent_directories=False):
+def search_parent(path, name):
+    """
+    Walk up the path to find the `name` (e.g. `.git`) directory.
+    :param dirname: Directory from which to start searching.
+    :param name: Directory to look for (e.g. `.git`).
+    """
+
+    while path:
+
+        if os.path.exists(os.path.join(path, name)):
+            return path
+
+        path, tail = os.path.split(path)
+
+        if not tail:
+            return None
+
+
+def iter_matching_entrypoints(path, entrypoint, config):
+
     trace("looking for ep", entrypoint, path)
-    if not search_parent_directories:
+
+    if config.search_parent_directories:
         for ep in iter_entry_points(entrypoint):
```

Comment:
unfortunately this loop has a mistake that can break discovery in some cases

for example if scm repositories of different scms for whatever reason are contained within each other,
then the wrong outer one may be found 

unfortunately its necessary to walk the potential roots one by one, and check each level for all entrypoints


i proposes a walk_potential_roots(root, search_parents) , that returns root if search_parents if false,

Reply:
Yeah I realised this as well 😴, I will correct for it!